src/Psl/Option/Option.php

```diff
@@ -0,0 +1,260 @@
+     *
+     * @param Option<T> $option
+     *
+     * @return Option<T>
+     */
+    public function or(Option $option): Option
+    {
+        if ($this->option !== null) {
+            return $this;
+        }
+
+        return $option;
+    }
+
+    /**
+     * Returns none if the option is none, otherwise calls `$predicate` with the wrapped value and returns:
+     *  - Option<T>::some() if `$predicate` returns true (where t is the wrapped value), and
+     *  - Option<T>::none() if `$predicate` returns false.
+     *
+     * @param (Closure(T): bool) $predicate
+     *
+     * @return Option<T>
+     */
+    public function filter(Closure $predicate): Option
+    {
+        if ($this->option !== null) {
+            return $predicate($this->option[0]) ? $this : none();
+        }
+
+        return $this;
+    }
+
+    /**
+     * Returns true if the option is a `Option<T>::some()` value containing the given value.
+     *
+     * @psalm-assert-if-true T $value
+     */
+    public function contains(mixed $value): bool
+    {
+        if ($this->option !== null) {
+            return $this->option[0] === $value;
+        }
+
+        return false;
+    }
+
+    /**
+     * Maps an `Option<T>` to `Option<Tu>` by applying a function to a contained value.
+     *
+     * @template Tu
+     *
+     * @param (Closure(T): Tu) $closure
+     *
+     * @return Option<Tu>
+     */
+    public function map(Closure $closure): Option
+    {
+        if ($this->option !== null) {
+            return some($closure($this->option[0]));
+        }
+
+        /** @var Option<Tu> */
+        return $this;
+    }
+
+    /**
+     * Applies a function to the contained value (if some),
+     * or returns `Option<Tu>::some()` with the provided `$default` value (if none).
+     *
+     * @note:   arguments passed to `Option::mapOr()` are eagerly evaluated;
+     *          if you are passing the result of a function call, it is recommended to use `Option::mapOrElse()`, which is lazily evaluated.
+     *
+     * @template Tu
+     *
+     * @param (Closure(T): Tu) $closure
+     * @param Tu $default
+     *
+     * @return Option<Tu>
+     */
+    public function mapOr(Closure $closure, mixed $default): Option
+    {
+        if ($this->option !== null) {
+            return some($closure($this->option[0]));
+        }
+
+        return some($default);
+    }
+
+    /**
+     * Applies a function to the contained value (if some),
+     * or computes a default function result (if none).
+     *
+     * @template Tu
+     *
+     * @param (Closure(T): Tu)  $closure
+     * @param (Closure(): Tu)   $else
+     *
+     * @return Option<Tu>
+     */
+    public function mapOrElse(Closure $closure, Closure $default): Option
+    {
+        if ($this->option !== null) {
+            return some($closure($this->option[0]));
+        }
+
+        return some($default());
+    }
+}
```

Comment:
a `proceed` function could be added as well, to make it all consistent with result and either?

Reply:
it looks like `unwrapOrElse()` has exactly the behavior of possible `proceed()`

Reply:
Not completely, proceed would have a callback for both some and  for none that resolve to a specific type

Reply:
@veewee `mapOrElse()` then

Reply:
process is mapOrElse, this follows the naming convention used in rust option type, not sure if we should use proceed, as in other instances ( result/either ), both proceed callbacks take a value, which is either `T`/`TError`, or `TRight`/`TLeft`, but here, the second callable doesnt' take an argument

Reply:
Not really, map returns an Option.
Proceed should be acombination of mapOrElse and unwrap.

Reply:
We can always add later if we find the need for this.